Allow for multiple dependencies/dependents from the same package

alalazo opened the pull request:
fixes #11983 
closes #16447

same as #21683 which I cannot reopen after a force push for some reason

This PR changes the internal representation of `Spec` to allow for multiple dependencies or dependents stemming from the same package. This change permits to represent cases which are frequent in cross compiled environments or to bootstrap compilers.

Modifications:
- [x] Substitute `DependencyMap` with `_EdgeMap`. The main differences are that the latter does not support direct item assignment and can be modified only through its API. It also provides a `select_by` method to query items.
- [x] Reworked a few public APIs of `Spec` to get list of dependencies or related edges.
- [x] Added unit tests to prevent regression on #11983 and prove the synthetic construction of specs with multiple deps from the same package. 

Since #22845 went in first, this PR reuses that format and thus it should not change hashes. What happens is that in the list of dependencies we may have the same package being present multiple times with different associated specs.

alalazo committed:
Spec: allow storing more than one dep from a single package

Introduce a new data structure, called _EdgeMap
that maps package names to list of DependencySpec
objects (edges in the DAG).

This data structure is used to track both
dependencies and dependents of a node. It
allows filtering the stored values by
parent or child name and by dependency types.

Unit tests added:
- Regression on 11983
- Synthetic construction of split dependency
- Synthetic construction of bootstrapping

alalazo committed:
Added comments to address review

alalazo committed:
Using an enum for "store_by"

alalazo committed:
Fix petsc failure in gitlab pipeline

alalazo committed:
Fix style checks

alalazo committed:
Fixed rebasing issues

alalazo committed:
Minor: unify turning a list result into a set

alalazo committed:
minor: fixed docstring

alalazo committed:
Refactor _get_dependency

alalazo committed:
Fix copy in case of multiple dependencies from same package

alalazo committed:
Make spack.graph work with mutiple specs from a single package

alalazo committed:
Add a check to avoid using splicing with multiple specs etc.

alalazo committed:
Skip test that might result in spurious failures

becker33 commented on lib/spack/spack/graph.py:
```diff
@@ -42,54 +42,73 @@
-    Return a list of dependency specs sorted topologically.  The spec
-    argument is not modified in the process.
 
+def topological_sort(spec, deptype='all'):
```
This function could use some comments

alalazo replied:
Should be fixed in 950f436

tgamblin commented on lib/spack/spack/dependency.py:
```diff
@@ -58,7 +58,7 @@ def canonical_deptype(deptype):
             raise ValueError(
                 'Invalid dependency types: %s' % ','.join(str(t) for t in bad))
-        return tuple(sorted(deptype))
+        return tuple(sorted(set(deptype)))
```
Is this needed?  Canonical deptype is called a lot and can be performance sensitive.

alalazo replied:
I would call this a bug in the current `develop` implementation. With this change:
```python
Spack version 0.17.1
Python 3.8.10, Linux x86_64
>>> import spack.dependency
>>> spack.dependency.canonical_deptype(['link', 'build', 'link'])
('build', 'link')
```
without this change:
```python
Spack version 0.17.1
Python 3.8.10, Linux x86_64
>>> import spack.dependency
>>> spack.dependency.canonical_deptype(['link', 'build', 'link'])
('build', 'link', 'link')
```
Also, if compared with https://github.com/spack/spack/pull/28673#issuecomment-1048593565 it doesn't seem to affect timing a lot. I obtain:
```console
$ time spack buildcache list
==> 0 cached builds.
==> You can query all available architectures with:
  spack buildcache list --allarch

real	1m7,731s
user	1m5,399s
sys	0m0,350s
```
when I remove `set` from the implementation.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -1016,7 +1140,9 @@ def __init__(self, spec, name, query_parameters):
         # Adding new attributes goes after super() call since the ObjectWrapper
         # resets __dict__ to behave like the passed object
-        self.token = spec, name, query_parameters
+        original_spec = getattr(spec, 'wrapped_obj', spec)
+        self.wrapped_obj = original_spec
+        self.token = original_spec, name, query_parameters
```
Should this logic be pushed into lang.ObjectWrapper? It seems like it would apply to any object that cares about it's id, so we could use `__getattr__` and `__getitem__` to make sure methods are called on the wrapped object.

alalazo replied:
Hmm, I need to think about that. I think `__getattr__` wouldn't work since the way we create a wrapper object is to fake the class name and share the `__dict__` object:

https://github.com/spack/spack/blob/c791ddc742474226bed94fcfadd5965148957918/lib/spack/llnl/util/lang.py#L742-L747

We can probably create though a better API to say that a method should be forwarded to the real object.

becker33 reviewed (commented):
Mostly LGTM

alalazo committed:
Fix for docs

alalazo committed:
Added comments to topological_sort

becker33 reviewed (dismissed):
I think this is ready to go, but I'm a little nervous that we could be missing something.

balay reviewed (dismissed):
Approving petsc/package.py change [checked a couple of builds - they go through fine]

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+        'run' not in dspec.deptypes,
+        'build' not in dspec.deptypes,
+        'test' not in dspec.deptypes
+    )
```
```suggestion
    return tuple(t not in dspec.deptypes for t in ("link", "run", "build", "test"))
```

alalazo replied:
Fixed in c19ef50

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+    """Represent a collection of edges (DependencySpec objects) in the DAG.
+
+    Objects of this class are used in Specs to track edges that are
+    outgoing towards direct dependencies, or edges that are ingoing
```
```suggestion
    outgoing towards direct dependencies, or edges that are incoming
```

alalazo replied:
Fixed in c19ef50

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+
+        return clone
+
+    def select_by(self, parent=None, child=None, dependency_types='all'):
```
`dependency_types` should be renamed to `deptypes`, because we call it that in every other place it's used as an argument.

alalazo replied:
I slightly disagree with that. We call them both `deptype` and `deptypes` and we had bugs because of the two similar spellings[^1]. Writing the full name in my opinion helps avoiding these kind of issues.

[^1]: Actually looking again at the code base there are more bugs than I caught in the past. I'll submit a separate PR to fix them. See for instance this call: https://github.com/spack/spack/blob/1ddad522a4b1527d337bfface0a62b26b8520347/lib/spack/spack/package.py#L1192
where `deptypes` gets forwarded to this other method:
https://github.com/spack/spack/blob/1ddad522a4b1527d337bfface0a62b26b8520347/lib/spack/spack/spec.py#L1350
that gets a `deptype` (without `s`) instead.

tgamblin replied:
I think we should standardize on `deptypes` here and fix the other occurrences in the separate PR

alalazo replied:
See https://github.com/spack/spack/issues/29397 and f99c721

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+
+        return clone
+
+    def select_by(self, parent=None, child=None, dependency_types='all'):
```
Nitpick: I think this would be better named `select`, as `_by` is redundant.

alalazo replied:
Fixed in c4c6ff6

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+        """Return a list of edges selected by dependency types.
+
+        If an edge has any of the dependency types passed as argument
+        then it is selected.
```
docstring is wrong -- this is filtered by deptype, parent, and child

alalazo replied:
Fixed in c19ef50

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+
+        dependency_types = dp.canonical_deptype(dependency_types)
+        # Start from all the edges we store
+        selected = [d for d in itertools.chain.from_iterable(self.values())]
```
Use a generator instead of building the whole list here, and same for the rest of the filters
```suggestion
        selected = (d for d in itertools.chain.from_iterable(self.values()))
```

alalazo replied:
Done in ac4a84e

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+                or not dep.deptypes
+            ]
+
+        return selected
```
Assuming all the above are now generators:
```suggestion
        return list(selected)
```

Does the result need to be a list though?

alalazo replied:
Done in ac4a84e 

> Does the result need to be a list though?

I think returning a list of edges is sensible, and probably the simplest choice. Do you have other data structures in mind for this?

tgamblin replied:
not really -- I think this is good.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+            selected = [
+                dep for dep in selected
+                if any(d in dependency_types for d in dep.deptypes)
+                or not dep.deptypes
```
By reflex, I would put this before the `any` as it's a simpler check, but my intuition about performance could be wrong.

alalazo replied:
I'll do it, doesn't seem to change performance for:
```console
$ spack buildcache list
```
though.

alalazo replied:
Done in ac4a84e

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,136 @@ def __str__(self):
+            current_list.append(edge)
+
+        # Keep the list sorted
+        current_list.sort(key=_sort_by_dep_types)
```
I worry that this and some of the spec comparisons above are expensive.  Have you looked at how this PR affects the performance of things like `spack buildcache list` on large build caches?  Can you compare to `develop` and see if there is any significant impact?

alalazo replied:
See profiling in https://github.com/spack/spack/pull/28673#issuecomment-1048686877

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
-        if dep is not None:
-            return dep
-        raise InvalidDependencyError(self.name, name)
+    def reset_dependencies(self):
```
Why not `clear_dependencies`?

alalazo replied:
Fixed in bea237b

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
+        """Trim the dependencies of this spec."""
+        self._dependencies.clear()
+
+    def reset_edges(self):
```
`clear_edges()`?

alalazo replied:
Fixed in 80bf04d

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
+        self._dependencies.clear()
+        self._dependents.clear()
+
+    def detach_node(self, deptype='all'):
```
`detach()`

alalazo replied:
Fixed in d7a1bff7

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
+                for edge in dependents_copy:
+                    if edge.parent.dag_hash() == key:
+                        continue
+                    dep._dependents.add(edge)
```
Why remove and re-add instead of just removing?

Also should this detach from dependents as well?

alalazo replied:
> Also should this detach from dependents as well?

This operation is only used in `Database._remove` and I kept the same semantic (remove any reference of a spec from dependencies before decrementing the refcount). 

> Why remove and re-add instead of just removing?

We might have items in the list that refer to other specs from the same package. Think of the same `zlib` used for different `hdf5` specs. Since it's a list it was more obvious to do it this way rather than trying to use `list.remove`. I guess I might use a list comprehension, but I don't expect benefits doing that.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
+        # WARNING: one edge from package "name"
+        deps = self.edges_to_dependencies(name=name)
+        err_msg = 'expected only 1 "{0}" dependency, but got {1}'
+        assert len(deps) == 1, err_msg.format(name, len(deps))
```
When will this fail, and should it be an assertion?  If this is an error that a user would see with the original concretizer, it should really be a `SpackError`, not an `assert`.

becker33 replied:
I think the only situation this fails is if you install something that has multiple dependencies of the same name, then switch back to the original concretizer, and then specifically request that earlier install by hash.

tgamblin replied:
I still think that's an error the user needs to see -- not an assert.

alalazo replied:
Fixed in c19ef50

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
+        return [
+            d.spec for d in
+            self._dependencies.select_by(child=name, dependency_types=deptype)
+        ]
```
```suggestion
        return [d.spec for d in self.edges_to_dependencies(name, deptype)]
```

alalazo replied:
Done in dd78582

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1143,34 +1272,114 @@ def _format_module_list(modules):
+        return [
+            d.parent for d in
+            self._dependents.select_by(parent=name, dependency_types=deptype)
+        ]
```
```suggestion
        return [d.parent for d in self.edges_from_dependents(name, deptype)]
```

alalazo replied:
Done in dd78582

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1280,10 +1496,12 @@ def root(self):
         Spack specs have a single root (the package being installed).
         """
+        # FIXME: In the case of multiple parents this property does not
+        # FIXME: make sense. Should we revisit the semantics?
```
I think we should probably get rid of this method but I am not sure exactly how.  It would be nice to add an assert, but it's very hard to do that without making the method really inefficient.  If we really need to make this efficient we can track roots on add, but that strikes me as overkill and this is only used in a few places.  I think we can just keep a note for now to remove this method later.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1448,22 +1666,25 @@ def return_val(dspec):
                 succ = lambda dspec: dspec.spec
             elif direction == 'parents':
-                where = self._dependents
+                edges = self._dependents
```
Can't these assign, e.g.:

```python
edges = self.edges_to_dependencies
```

```python
edges = self.edges_from_dependents
```

alalazo replied:
Done in 3315459

tgamblin replied:
Looks way better now!

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1448,22 +1666,25 @@ def return_val(dspec):
-                    continue
+            for name in sorted(edges):
+                edges_from_name = edges[name]
+                for dspec in edges_from_name:
```
Then, here, do:

```python
for dspec in sorted(edges(), key=...):
```

That would simplify the loop nest.

alalazo replied:
Done in 3315459

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1684,17 +1905,17 @@ def to_node_dict(self, hash=ht.dag_hash):
-                deps_list.append(syaml.syaml_dict([name_tuple,
-                                                   hash_tuple,
-                                                   type_tuple]))
+                for dspec in edges_for_name:
```
I am not sure that it makes sense to write this fully generally, as we can *still* never have two direct dependencies with the same name.  I think right here, we should really be asserting that there is only one thing in the list, then taking the first element, instead of adding a loop nest.

I am having a hard time thinking of when we would generalize enough to have multiple direct dependencies *of the same name*, and I worry that we've generalized the interface for the uncommon case of traversing dependents, instead of keeping it simple for the much more common case of traversing dependencies.  I worry that will lead to weird errors.

tgamblin replied:
Resolving this per later discussion

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -2257,8 +2478,14 @@ def _concretize_helper(self, concretizer, presets=None, visited=None):
+            # WARNING: algorithm we don't allow multiple nodes from
+            # WARNING: the same package in a DAG, here we hard-code
+            # WARNING: using index 0 i.e. we assume that we have only
+            # WARNING: one edge from package "name"
```
When will we ever (even now) have two *dependencies* with the same name?

tgamblin replied:
resolving per discussion points below.

alalazo replied:
Closing this, see https://github.com/spack/spack/pull/28673#issuecomment-1048557787

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -3229,12 +3456,19 @@ def _constrain_dependencies(self, other):
+                # WARNING: one edge from package "name"
+                edges_from_name = self._dependencies[name]
+                changed |= edges_from_name[0].update_deptypes(
+                    other._dependencies[name][0].deptypes)
```
Same question as above: when will we ever have two *dependencies* with the same name?

tgamblin replied:
resolving per discussion points below.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -3554,22 +3788,25 @@ def _dup(self, other, deps=True, cleardeps=True, caches=None):
-                                          root=False):
-            if (dspec.deptypes and
-                not any(d in deptypes for d in dspec.deptypes)):
+        def spid(spec):
+            return id(spec)
```
Why's this needed?  I think it's clearer with `id`

alalazo replied:
It's a minor point, but what we need is something that is unique by spec during the "lifetime" of a copy. Using an indirect name means I can change it in a single line instead of having to go through multiple places where this is used.

tgamblin replied:
ok fair enough.  I think this will always be `id` though.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1878,12 +1879,12 @@ def external_spec_selected(self, pkg, idx):
-                self._specs[dep], (type,))
+        dependencies = self._specs[pkg].edges_to_dependencies(name=dep)
+        if not dependencies:
+            self._specs[pkg].add_dependency_edge(self._specs[dep], (type,))
```
I think this interface ends up being confusing for dependencies.  There will still *never* be two of these.  I think we should avoid using a symmetric interface here -- `edges_to_dependencies()` makes sense to me but `edges_to_dependencies(name)` introduces a bunch of extra logic that we usually do not need.  Can we add `edge_to_dependency(name)` that returns only one thing, to simplify?

A singular `edge_from_dependent()` routine should probably *not* exist, as you could have multiple of them, and I think the client code should have to think about that.

`dependencies` just throws off my thinking here because there will never be more than one of these *for a given node*.

becker33 replied:
See my comment below, but as discussed on our meeting today this is incorrect for future work planned.

tgamblin replied:
resolving per discussion points below.

tgamblin commented on lib/spack/spack/graph.py:
```diff
@@ -42,54 +42,84 @@
+import llnl.util.tty.color
 
-from spack.dependency import all_deptypes, canonical_deptype
+import spack.dependency
```
```python
import spack.dependency as dp
```

would make this a bit more readable

alalazo replied:
Hmm, not sure. Using the fully qualified name I know from where a name is pulled without looking at a context that might be hundreds of lines above.

tgamblin reviewed (changes requested):
Ok I have two main things here:

1. Can you check the performance of this implementation?  With something like `spack buildcache list` where we have to construct and sort a lot of specs?  I am worried that the edge structures are going to be too complicated and that we'll run into issues with large build caches.  If there's no noticeable slowdown then I think that's fine, but it would be good to know b/c this is a major change to a core data structure.  I made a few comments inline to this effect.

2. ~~The interface for accessing dependencies here seems awkward in places.  I get why there are both `edges_from _dependents(name)` and `edges_to _dependencies(name)`, for symmetry, but essentially every place we ask for a dependency by name, there is going to be one element in that list.  The dependencies case (where there will only be one) is far more common than the dependents case (where we have to think about multiple).  So I think the interface should *not* be the same for these.~~

    ~~I think there should be a way to get the single edge with, e.g., `edge_to_dependency(name)` (which should fail with an assertion error if there are multiple), and `edges_from _dependents` should probably *not* take a `name` argument.  I don't think this leaks out of `spec.py` *too* much -- the dicts with list values are mostly hidden in the details of `_EdgeMap`, but there were places where I found the code harder to follow and had to think about the lists in `_EdgeMap` when they did not matter at all - that was hard to read.~~

becker33 commented:
For Todd's point (2), we need the dependency edge information for when we split build dependencies for cross-compilation. It's not possible yet, and it won't be in the first separate concretization of build deps PR, but we will need that eventually. I think that's a good enough reason to keep this now, since we'll need it in the long run.

alalazo commented:
> we need the dependency edge information for when we split build dependencies for cross-compilation.

I agree with @becker33 That is one of the point why we have this PR, to allow splitting edges for the same dependency if need be. A prominent example would be `python` being a `(build,run)` dependency in a cross-compilation environment.

Also, the common case - the one where we have a single dependency - should be covered by `__getitem__`:
```python
dep = root[dep_name]
```

tgamblin commented:
Ok, yes you guys are right -- *when* we properly handle cross-compilation and can split a dependency node between build and link/run environments, we'll care about this. So for point 2 above, yeah, let's keep the interface symmetric.

I guess right now, `root[dep_name]` will give you the `link` dependency if there happen to be two with the same name, which I guess is the right thing to do here.  I guess we can make the indexing fancier if we want later -- e.g., `root[dep_name, 'build']` if we really have to refine it.

tgamblin commented:
Ok updated the review and resolved the points around multiple edges to dependencies.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1878,12 +1879,12 @@ def external_spec_selected(self, pkg, idx):
-        if not dependency:
-            self._specs[pkg]._add_dependency(
-                self._specs[dep], (type,))
+        dependencies = self._specs[pkg].edges_to_dependencies(name=dep)
```
For now, to make this clearer, I'd suggest something like:

```suggestion
        dependencies = self._specs[pkg].edges_to_dependencies(name=dep)
        assert len(dependencies) == 1, "Current solver does not handle multiple dependency edges of the same name"
        dependency = dependencies[0] if dependencies else None
```

alalazo replied:
Done in 5d77b88

alalazo commented:
From a fresh clone of Spack, setting up the instance with these commands:
```console
$ spack solve zlib
$ spack mirror add E4S https://cache.e4s.io
$ spack buildcache keys -it
==> Fetching https://cache.e4s.io/build_cache/_pgp/25645FA2B218FE55B4EF649E4345F04B40005581.pub
gpg: key 4345F04B40005581: public key "University of Oregon - E4S" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: inserting ownertrust of 6
```
and running on:

* **Spack:** 0.17.1-1343-455fd15e86
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

This is the timing I obtain on this branch:
```
$ time spack buildcache list
==> 0 cached builds.
==> You can query all available architectures with:
  spack buildcache list --allarch

real	1m8,325s
user	1m5,968s
sys	0m0,399s

$ time spack buildcache list --allarch
[ ... ]
real	1m13,419s
user	1m10,911s
sys	0m0,453s
```
On `develop`:
```
$ time spack buildcache list
==> 0 cached builds.
==> You can query all available architectures with:
  spack buildcache list --allarch

real	0m51,076s
user	0m38,516s
sys	0m0,918s

$ time spack buildcache list --allarch
[ ... ]
real	0m52,793s
user	0m50,168s
sys	0m0,424s
```
so it seems there's a ~15-20% increase in time for this call.

tgamblin commented:
Can you profile it with pyinstrument?  When I've looked at this in the past it was very sensitive to canonicalizing dependencies in too many places (as that happens a lot as part of traversal).  I'm suspicious of that and maybe the new list structure -- but moreso the canonicalizing.

tgamblin commented:
Also another way to benchmark this that's faster than waiting for `buildcache list` is to concretize a large spec, save it as json, then run pyinstrument on a script that just loads it and traverses it a hundred (maybe a few hundred?) times.

I tend to run pyinstrument like this, to show everything:

```python
python3 -m pyinstrument --show-all --hide zzz $(which spack) help
```

`--hide zzz` overrides the somewhat annoying default hide settings it has -- which apparently are not overridden by `--show-all`

alalazo committed:
Address a few minor points raised in review for spec.py

Fixed docstrings, raise instead of assert,
don't inline a for loop to create a tuple

alalazo commented:
With c19ef50 I have this profiling [buildcache_list.spy.txt](https://github.com/spack/spack/files/8123877/buildcache_list.spy.txt) obtained with:
```console
$ py-spy record -f speedscope -o buildcache_list.spy -s -n $(which spack) buildcache list
```
It can be seen at https://www.speedscope.app/ It seems we spend a lot of time (for what this PR is concerned) in the comparisons at:

https://github.com/spack/spack/blob/c19ef50986dbc8b217426907f8d85149cc3d5201/lib/spack/spack/spec.py#L826

If I read the profiling data correctly canonicalizing doesn't have a major impact for the command above.

alalazo committed:
Rename select_by -> select

alalazo committed:
Rename reset_dependencies -> clear_dependencies

alalazo committed:
Rename reset_edges -> clear_edges

alalazo committed:
Rename detach_node -> detach

alalazo committed:
Make selection filters generators, swap condition evaluation in OR

alalazo committed:
Use public API instead of private attributes for dependencies/dependents

alalazo committed:
Avoid explicitly nested loop in traverse edges

alalazo committed:
Add assertion in asp.py to avoid having 2 deps of the same name

alalazo commented:
@tgamblin I think all the points have been either fixed or replied to. The most important discussion still open is possibly the one about performance with profiling data in this https://github.com/spack/spack/pull/28673#issuecomment-1048686877. I'll try to have another look at this tomorrow, but let me know how do you want to proceed with that ~20% increase in time.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1248,18 +1449,25 @@ def _add_dependency(self, spec, deptypes):
-        spec._dependents[self.name] = dspec
+        self.add_dependency_edge(spec, deptypes)
+
+    def add_dependency_edge(self, dependency_spec, deptype):
```
Can this just be called `add_dependency`?  Rationale:

1. To some extent, the edges are an abstract concept to the user of `Spec`, though we do have methods that can iterate over edges.
2. But, this function doesn't take an edge.  It take a `Spec` -- i.e., the actual dependency.  So I think `edge` here can be confusing to the caller.

tgamblin replied:
I'm just now noticing that `_add_dependency` still exists and just calls this.  I think these should be consolidated to `add_dependency`

alalazo replied:
> I'm just now noticing that _add_dependency still exists and just calls this. I think these should be consolidated to add_dependency

`_add_dependency` is there just to enforce the constraint that we can't have two dependencies with the same name at the moment. It will be removed as soon as we arrive at separate concretization of build deps. 

For the naming of `add_dependency_edge`  I have no preference. My intention with having `edge` in the name was to give a hint that:
1. The dependency spec (the DAG node node) will not be changed in any way by this call
2. All the properties that are added (deptypes and in the future virtuals etc.) are associated with an edge

alalazo replied:
@tgamblin This renaming is still missing. If the Pr seems good enough I can follow with a smaller PR that does just the renaming. Otherwise, what should be the new names? I was thinking:
- `_add_dependency` -> `legacy_add_dependency` (this will be removed when we add separate concretization of build deps)
- `add_dependency_edge` -> `add_dependency`

tgamblin commented:
So on the performance issue: on `develop` on my mac, `spack build cache list` takes 45s, and on this branch it takes 76s.  That is 68% slower so I think we really need to address it.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +767,133 @@ def __str__(self):
+        # If there's something in the list, check if we need to update an
+        # already existing entry.
+        for dep in current_list:
+            if edge.spec == dep.spec and edge.parent == dep.parent:
```
Jut as a preliminary look into this -- I think the solution here is to *not* do this duplicate checking (at least not with a full spec compare) on insert.  We did not do it before -- though the data structure didn't allow this much flexibility.

Just using `spack find` seems not to add duplicates on my small DB, and the CLI spec parser doesn't allow it.  When I do a `buildcache list`, it looks like the vast majority of adds are not duplicates, but there are a few (29 out of tens of thousands) that are.  In all cases they aren't adding new deptypes or anything -- they're adding the same deptypes that are already there.  I'm going to look at usages of `add_dependency_edge` and `_add_dependency` to see if we can eliminate this checking.

becker33 replied:
If we just store a separate edge for each dep type between the two specs, will that hurt us anywhere else? We'll have to change how we check the edge list for dependencies, but that should be doable.

And then we can just check name and type here, and if the name/type combination already exists, we can raise.

tgamblin replied:
That doesn’t work for dependents. You can have the same name/type multiple times.

alalazo replied:
There's also the fact that here we are really talking about edge attributes. Right now we only have dependency types but in the future there will be more (virtuals provided, use variants etc.) - so one edge per attribute value I think won't work.

alalazo replied:
Leaving a note on the idea to use `id` instead of `__eq__` for this check. When we compare abstract specs using the check above with `__eq__` we have:
```python
Spack version 0.17.1
Python 3.8.10, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec('root')
>>> d1 = spack.spec.Spec('child@1.0')
>>> d2 = spack.spec.Spec('child@1.0')
>>> s.add_dependency_edge(d1, deptype=('link',))
>>> s.add_dependency_edge(d2, deptype=('build',))
>>> str(s)
'root ^child@1.0'
```
i.e. the two calls unify d1 and d2 into a single `(build,link)` edge. This seems sensible to me. Using `id` will surely be faster, but opens for situations like the following on abstract specs:
```python
>>> import spack.spec
>>> s = spack.spec.Spec('root')
>>> d1 = spack.spec.Spec('child@1.0')
>>> d2 = spack.spec.Spec('child@1.0')
>>> s.add_dependency_edge(d1, deptype=('link',))
>>> s.add_dependency_edge(d2, deptype=('link',))
>>> str(s)
'root ^child@1.0 ^child@1.0'
```
where adding even the same dependency type result in a duplicate edge because `d1` and `d2` are not the same object.

alalazo replied:
Considering that `id(s1) == id(s2)` implies both:
- `s1 == s2`
- `s1.*_hash() == s2.*_hash()`, if the specs are concrete

is it an idea we can consider to keep them both and compare by `id` first (fast method) and by hash/equality later, i.e. if `id` are not the same?

alalazo replied:
It seems the idea above is already implemented, so taking note of that here:

https://github.com/spack/spack/blob/6934703ffbae9b475b66c6321a1fb06c500105fa/lib/spack/llnl/util/lang.py#L436-L439

After a few DMs exchange with @tgamblin the idea is:
1. Remove the double comparison from line 826
2. Have `Spec.add_dependency` (see https://github.com/spack/spack/pull/28673#discussion_r820165922) take care or vetting input specs if needed

alalazo replied:
See df30e6c 085dcf8 and 43bd292

alalazo committed:
Move edge checks from _EdgeMap.add to Spec

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -1458,6 +1442,13 @@ def add_dependency_edge(self, dependency_spec, deptype):
         """
+        # Check if we need to update edges that are already present
+        selected = self._dependencies.select(child=dependency_spec.name)
+        for edge in selected:
+            if id(dependency_spec) == id(edge.spec):
+                edge.add_type(deptype)
+                return
```
@tgamblin ~I still need to update this part, to do the symmetric operation on the parent It seems though we have no test that detected this.~ Ok, it turns out this is correct, since we push the same object into two dictionaries, so when we update the dependency types the other reference sees the update too.

alalazo committed:
Add two unit-tests to validate the addition of edges

alalazo committed:
Rename "dependency_types" to "deptypes"

alalazo committed:
Reduce the number of calls to dependency.canonical_deptype

alalazo commented:
@tgamblin From what I can see locally the timing of this branch is now more in line with `develop`. On 43bd292 I got:
```console
$ time spack buildcache list
==> 0 cached builds.
==> You can query all available architectures with:
  spack buildcache list --allarch

real	0m47,115s
user	0m44,382s
sys	0m0,356s

$ time spack buildcache list --allarch
[ ... ]
real	0m50,916s
user	0m48,500s
sys	0m0,429s
```
while on `develop:
```console
$ time spack buildcache list
==> 0 cached builds.
==> You can query all available architectures with:
  spack buildcache list --allarch

real	0m46,886s
user	0m44,530s
sys	0m0,380s

$ time spack buildcache list --allarch
[ ... ]
real	0m44,941s
user	0m42,481s
sys	0m0,469s
```
For the record, I am using:

* **Spack:** 0.17.1-1492-43bd292e13
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

alalazo committed:
Remove a redundant sort operation

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -766,13 +769,119 @@ def __str__(self):
+        key = edge.spec.name if self.store_by_child else edge.parent.name
+        current_list = self.edges.setdefault(key, [])
+        current_list.append(edge)
+        current_list.sort(key=_sort_by_dep_types)
```
Note that while this guarantees deterministic dependency traversal order, it doesn't guarantee a deterministic dependent traversal order for the same spec.  I do not think that matters currently (we don't hash dependent traversals) but it's worth remembering.

tgamblin commented on lib/spack/spack/test/spec_dag.py:
```diff
@@ -989,3 +1021,99 @@ def test_spec_tree_respect_deptypes(self):
+    c2 = Spec('b@2.0').concretized()
+
+    p.add_dependency_edge(c1, deptype=c1_deptypes)
+    with pytest.raises(spack.error.SpackError):
```
I verified that this raises the right exception but for the future: this test should probably not have the parent *and* the dependencies as `b`.  If we added cycle checking or something I could see this error getting confused with another.

This is true for a few tests above too

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1248,18 +1440,43 @@ def _add_dependency(self, spec, deptypes):
+                raise spack.error.SpecError(msg.format(edge, deptype))
+
+        for edge in selected:
+            if id(dependency_spec) == id(edge.spec):
```
this should probably be shortened to `is` but I think we can leave it for now.

tgamblin reviewed (approved):
I have a few really minor nitpicks above but none of them should impede getting this merged -- they can be follow-ons.  LGTM!  The new tests and minimal performance hit are great.

sethrj commented:
@alalazo  I think this change is responsible for https://github.com/spack/spack/issues/29478: installing the exact same python spec with the new spack core fails because `link_deps = spec.dependencies('link')` seems to evaluate to an empty list now. Is this behavior change expected? What can be done to fix it?

alalazo commented:
@sethrj Changes to fix the package:
```python
spec.dependencies(deptype='link')
```
As it is now it's looking for a dependency named `link`. Let me submit a PR.
